File: backend/app/services/indicador_service.py
from datetime import datetime, timezone
from sqlalchemy.orm import Session, joinedload, noload
from decimal import Decimal
from uuid import UUID
from fastapi import HTTPException, status

from ..models.calidad import Indicador, MedicionIndicador
from ..utils.audit import registrar_auditoria


TIPOS_INDICADOR = {"eficacia", "eficiencia", "cumplimiento"}
ESTADOS_INDICADOR = {"borrador", "pendiente_aprobacion", "aprobado", "rechazado"}


def _ahora():
    return datetime.now(timezone.utc)


class IndicadorService:
    def __init__(self, db: Session):
        self.db = db
        try:
            from ..db.ensure_schema import asegurar_esquema_calidad
            asegurar_esquema_calidad()
        except Exception:
            pass

    def _query(self, con_mediciones: bool = True):
        opciones = [
            joinedload(Indicador.proceso),
            joinedload(Indicador.responsable_medicion),
            joinedload(Indicador.creador),
            joinedload(Indicador.revisador),
            joinedload(Indicador.aprobador),
        ]
        if con_mediciones:
            opciones.append(joinedload(Indicador.mediciones).joinedload(MedicionIndicador.registrador))
        else:
            opciones.append(noload(Indicador.mediciones))
        return self.db.query(Indicador).options(*opciones)

    def obtener(self, indicador_id: UUID) -> Indicador:
        try:
            indicador = self._query().filter(Indicador.id == indicador_id).first()
        except Exception:
            self.db.rollback()
            indicador = self._query(con_mediciones=False).filter(Indicador.id == indicador_id).first()
        if not indicador:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Indicador no encontrado")
        return indicador

    def _aplicar_filtros(self, query, proceso_id: UUID = None, activo: bool = None, tipo_indicador: str = None):
        if proceso_id:
            query = query.filter(Indicador.proceso_id == proceso_id)
        if activo is not None:
            query = query.filter(Indicador.activo == activo)
        if tipo_indicador:
            tipo = tipo_indicador.strip().lower()
            if tipo not in TIPOS_INDICADOR:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Tipo inválido. Use: {', '.join(sorted(TIPOS_INDICADOR))}",
                )
            query = query.filter(Indicador.tipo_indicador == tipo)
        return query

    def listar(self, proceso_id: UUID = None, activo: bool = None, tipo_indicador: str = None, skip: int = 0, limit: int = 200):
        query = self._aplicar_filtros(self._query(), proceso_id, activo, tipo_indicador)
        try:
            return query.order_by(Indicador.codigo.asc()).offset(skip).limit(limit).all()
        except Exception:
            self.db.rollback()
            query = self._aplicar_filtros(self._query(con_mediciones=False), proceso_id, activo, tipo_indicador)
            return query.order_by(Indicador.codigo.asc()).offset(skip).limit(limit).all()

    def registrar_medicion(self, indicador_id: UUID, data: dict, usuario_id: UUID) -> MedicionIndicador:
        indicador = self.obtener(indicador_id)

        meta = data.get("meta")
        if meta is None:
            meta = indicador.meta

        cumple_meta = None
        if meta is not None:
            cumple_meta = Decimal(str(data["valor"])) >= Decimal(str(meta))

        medicion = MedicionIndicador(
            indicador_id=indicador_id,
            periodo=data["periodo"],
            valor=data["valor"],
            meta=meta,
            cumple_meta=cumple_meta,
            observaciones=data.get("observaciones"),
            registrado_por=usuario_id,
            creado_por=usuario_id,
        )
        self.db.add(medicion)

        indicador.revisado_por = usuario_id
        indicador.fecha_revision = _ahora()
        if indicador.estado == "aprobado":
            indicador.estado = "pendiente_aprobacion"
        elif indicador.estado == "rechazado":
            indicador.estado = "borrador"

        self.db.flush()

        registrar_auditoria(
            self.db,
            tabla="mediciones_indicador",
            registro_id=medicion.id,
            accion="CREATE",
            usuario_id=usuario_id,
            cambios={"indicador_id": str(indicador_id), **data, "cumple_meta": cumple_meta},
        )
        self.db.commit()
        self.db.refresh(medicion)
        return medicion

    def historial(self, indicador_id: UUID):
        self.obtener(indicador_id)
        try:
            return (
                self.db.query(MedicionIndicador)
                .options(joinedload(MedicionIndicador.registrador))
                .filter(MedicionIndicador.indicador_id == indicador_id)
                .order_by(MedicionIndicador.periodo.asc(), MedicionIndicador.creado_en.asc())
                .all()
            )
        except Exception:
            self.db.rollback()
            return (
                self.db.query(MedicionIndicador)
                .filter(MedicionIndicador.indicador_id == indicador_id)
                .order_by(MedicionIndicador.periodo.asc(), MedicionIndicador.creado_en.asc())
                .all()
            )

    def tendencia(self, indicador_id: UUID) -> dict:
        mediciones = self.historial(indicador_id)
        if not mediciones:
            return {
                "indicador_id": indicador_id,
                "total_mediciones": 0,
                "promedio": Decimal("0"),
                "ultimo_valor": None,
                "ultimo_periodo": None,
                "tendencia": "sin_datos",
            }

        valores = [Decimal(str(m.valor)) for m in mediciones if m.valor is not None]
        if not valores:
            return {
                "indicador_id": indicador_id,
                "total_mediciones": len(mediciones),
                "promedio": Decimal("0"),
                "ultimo_valor": None,
                "ultimo_periodo": mediciones[-1].periodo if mediciones else None,
                "tendencia": "sin_datos",
            }

        promedio = sum(valores) / Decimal(len(valores))

        tendencia = "estable"
        if len(valores) >= 2:
            if valores[-1] > valores[-2]:
                tendencia = "subiendo"
            elif valores[-1] < valores[-2]:
                tendencia = "bajando"

        ultima = mediciones[-1]
        ultimo_val = Decimal(str(ultima.valor)) if ultima.valor is not None else None
        return {
            "indicador_id": indicador_id,
            "total_mediciones": len(mediciones),
            "promedio": promedio.quantize(Decimal("0.01")),
            "ultimo_valor": ultimo_val,
            "ultimo_periodo": ultima.periodo,
            "tendencia": tendencia,
        }

    def solicitar_aprobacion(self, indicador_id: UUID, usuario_id: UUID) -> Indicador:
        indicador = self.obtener(indicador_id)
        if indicador.estado == "aprobado":
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="El indicador ya está aprobado")
        mediciones = self.historial(indicador_id)
        if not mediciones:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Registre al menos una medición antes de solicitar aprobación",
            )
        indicador.estado = "pendiente_aprobacion"
        indicador.revisado_por = usuario_id
        indicador.fecha_revision = _ahora()
        try:
            registrar_auditoria(
                self.db,
                tabla="indicadores",
                registro_id=indicador.id,
                accion="UPDATE",
                usuario_id=usuario_id,
                cambios={"estado": "pendiente_aprobacion"},
            )
        except Exception:
            pass
        self.db.commit()
        return self.obtener(indicador_id)

    def aprobar(self, indicador_id: UUID, usuario_id: UUID, observacion: str | None = None) -> Indicador:
        indicador = self.obtener(indicador_id)
        if indicador.estado not in {"borrador", "pendiente_aprobacion", "rechazado"}:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="El indicador no está pendiente de aprobación")
        if indicador.creado_por and str(indicador.creado_por) == str(usuario_id):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Quien elabora el indicador no puede aprobarlo. Debe aprobarlo otra persona.",
            )
        mediciones = self.historial(indicador_id)
        if not mediciones:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="No se puede aprobar un indicador sin mediciones",
            )
        indicador.estado = "aprobado"
        indicador.aprobado_por = usuario_id
        indicador.fecha_aprobacion = _ahora()
        indicador.observacion_aprobacion = observacion
        if not indicador.revisado_por:
            indicador.revisado_por = indicador.responsable_medicion_id or usuario_id
            indicador.fecha_revision = indicador.fecha_revision or _ahora()
        try:
            registrar_auditoria(
                self.db,
                tabla="indicadores",
                registro_id=indicador.id,
                accion="UPDATE",
                usuario_id=usuario_id,
                cambios={"estado": "aprobado", "aprobado_por": str(usuario_id)},
            )
        except Exception:
            pass
        self.db.commit()
        return self.obtener(indicador_id)

    def rechazar(self, indicador_id: UUID, usuario_id: UUID, observacion: str | None = None) -> Indicador:
        indicador = self.obtener(indicador_id)
        if indicador.estado not in {"borrador", "pendiente_aprobacion"}:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="El indicador no está pendiente de aprobación")
        if indicador.creado_por and str(indicador.creado_por) == str(usuario_id):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Quien elabora el indicador no puede rechazarlo. Debe decidirlo otra persona.",
            )
        if not observacion or not observacion.strip():
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Indique el motivo del rechazo")
        indicador.estado = "rechazado"
        indicador.aprobado_por = None
        indicador.fecha_aprobacion = None
        indicador.observacion_aprobacion = observacion.strip()
        try:
            registrar_auditoria(
                self.db,
                tabla="indicadores",
                registro_id=indicador.id,
                accion="UPDATE",
                usuario_id=usuario_id,
                cambios={"estado": "rechazado", "observacion": observacion.strip()},
            )
        except Exception:
            pass
        self.db.commit()
        return self.obtener(indicador_id)

File: frontend/src/services/indicador.service.ts
import apiClient from "@/lib/api";

export type TipoIndicador = "eficacia" | "eficiencia" | "cumplimiento";
export type EstadoIndicador = "borrador" | "pendiente_aprobacion" | "aprobado" | "rechazado" | string;

export interface UsuarioIndicador {
  id: string;
  nombre: string;
  segundoNombre?: string;
  segundo_nombre?: string;
  primerApellido?: string;
  primer_apellido?: string;
  segundoApellido?: string;
  segundo_apellido?: string;
}

export interface MedicionIndicador {
  id: string;
  indicador_id: string;
  periodo: string;
  valor: number;
  meta?: number | null;
  cumple_meta?: boolean | null;
  observaciones?: string | null;
  registrado_por?: string | null;
  creado_en: string;
  actualizado_en: string;
  registrador?: UsuarioIndicador | null;
}

export interface Indicador {
  id: string;
  proceso_id: string;
  codigo: string;
  nombre: string;
  descripcion?: string;
  formula?: string;
  unidad_medida?: string;
  meta?: number;
  frecuencia_medicion: string;
  responsable_medicion_id?: string;
  tipo_indicador?: TipoIndicador | string;
  estado?: EstadoIndicador;
  activo: boolean;
  creado_por?: string;
  revisado_por?: string | null;
  fecha_revision?: string | null;
  aprobado_por?: string | null;
  fecha_aprobacion?: string | null;
  observacion_aprobacion?: string | null;
  creado_en: string;
  actualizado_en: string;
  proceso?: {
    id: string;
    codigo?: string;
    nombre: string;
  };
  responsable?: UsuarioIndicador | null;
  creador?: UsuarioIndicador | null;
  revisador?: UsuarioIndicador | null;
  aprobador?: UsuarioIndicador | null;
  ultima_medicion?: MedicionIndicador | null;
}

export interface TendenciaIndicador {
  indicador_id: string;
  total_mediciones: number;
  promedio: number;
  ultimo_valor?: number | null;
  ultimo_periodo?: string | null;
  tendencia: "subiendo" | "bajando" | "estable" | "sin_datos" | string;
}

export const indicadorService = {
  async getAll(filters?: {
    proceso_id?: string;
    activo?: boolean;
    tipo_indicador?: TipoIndicador | string;
  }): Promise<Indicador[]> {
    const response = await apiClient.get("/indicadores", {
      params: {
        proceso_id: filters?.proceso_id,
        activo: filters?.activo,
        tipo_indicador: filters?.tipo_indicador,
      },
    });
    return response.data;
  },

  async getActivos(): Promise<Indicador[]> {
    return this.getAll({ activo: true });
  },

  async getById(id: string): Promise<Indicador> {
    const response = await apiClient.get(`/indicadores/${id}`);
    return response.data;
  },

  async create(data: Partial<Indicador>): Promise<Indicador> {
    const response = await apiClient.post("/indicadores", data);
    return response.data;
  },

  async update(id: string, data: Partial<Indicador>): Promise<Indicador> {
    const response = await apiClient.put(`/indicadores/${id}`, data);
    return response.data;
  },

  async delete(id: string): Promise<void> {
    await apiClient.delete(`/indicadores/${id}`);
  },

  async registrarMedicion(id: string, data: {
    periodo: string;
    valor: number;
    meta?: number;
    observaciones?: string;
  }): Promise<MedicionIndicador> {
    const response = await apiClient.post(`/indicadores/${id}/mediciones`, data);
    return response.data;
  },

  async getMediciones(id: string): Promise<MedicionIndicador[]> {
    const response = await apiClient.get(`/indicadores/${id}/mediciones`);
    return response.data;
  },

  async getTendencia(id: string): Promise<TendenciaIndicador> {
    const response = await apiClient.get(`/indicadores/${id}/tendencia`);
    return response.data;
  },

  async solicitarAprobacion(id: string): Promise<Indicador> {
    const response = await apiClient.post(`/indicadores/${id}/solicitar-aprobacion`);
    return response.data;
  },

  async aprobar(id: string, observacion?: string): Promise<Indicador> {
    const response = await apiClient.post(`/indicadores/${id}/aprobar`, { observacion: observacion || null });
    return response.data;
  },

  async rechazar(id: string, observacion: string): Promise<Indicador> {
    const response = await apiClient.post(`/indicadores/${id}/rechazar`, { observacion });
    return response.data;
  },
};
